src/core/exec-invoke.c

```diff
@@ -2223,6 +2223,53 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
         return 0;
 }
 
+static int setup_private_users_child(int unshare_ready_fd, const char *uid_map, const char *gid_map, bool allow_setgroups) {
+        /* Child process, running in the original user namespace. Let's update the parent's UID/GID map from
+         * here, after the parent opened its own user namespace. */
+
+        pid_t ppid = getppid();
+
+        /* Wait until the parent unshared the user namespace */
+        uint64_t c;
+        if (read(unshare_ready_fd, &c, sizeof(c)) < 0)
+                return log_debug_errno(errno, "Failed to read from signaling eventfd: %m");
+
+        /* Disable the setgroups() system call in the child user namespace, for good, unless PrivateUsers=full
+         * and using the system service manager. */
+        const char *a = procfs_file_alloca(ppid, "setgroups");
+        _cleanup_close_ int fd = open(a, O_WRONLY|O_CLOEXEC);
+        if (fd < 0)
+                return log_debug_errno(errno, "Failed to open %s: %m", a);
+
+        const char *setgroups = allow_setgroups ? "allow\n" : "deny\n";
+        if (write(fd, setgroups, strlen(setgroups)) < 0)
+                return log_debug_errno(errno, "Failed to write '%s' to %s: %m", setgroups, a);
```

Comment:
write_string_file

Reply:
Yes, I just moved that code block so I tried to touch as little as possible